File: superstitioApi/java/superstitioapi/powers/barIndepend/HasBarRenderOnCreature_Power.java
package superstitioapi.powers.barIndepend;

import com.megacrit.cardcrawl.helpers.Hitbox;
import com.megacrit.cardcrawl.powers.AbstractPower;

import java.util.function.BiFunction;
import java.util.function.Supplier;

public interface HasBarRenderOnCreature_Power extends HasBarRenderOnCreature {

    AbstractPower getSelf();

    default BiFunction<Supplier<Hitbox>, HasBarRenderOnCreature, ? extends RenderOnThing> makeNewBarRenderOnCreature() {
        return BarRenderOnThing::new;
    }

    @Override
    default String uuidOfSelf() {
        return getSelf().ID;
    }

    @Override
    default String uuidPointTo() {
        return HasBarRenderOnCreature.super.uuidPointTo() + ":" + this.getSelf().owner.id;
    }

    @Override
    default int getAmountForDraw() {
        return getSelf().amount;
    }

    int maxBarAmount();

    @Override
    default String getName() {
        return getSelf().name;
    }

    @Override
    default String getDescription() {
        return getSelf().description;
    }

    @Override
    default Hitbox getBarRenderHitBox() {
        return getSelf().owner.hb;
    }
}

File: superstitioApi/java/superstitioapi/powers/barIndepend/HasBarRenderOnCreature_Power.kt
package superstitioapi.powers.barIndepend

import com.megacrit.cardcrawl.characters.AbstractPlayer
import com.megacrit.cardcrawl.helpers.Hitbox
import com.megacrit.cardcrawl.powers.AbstractPower
import java.util.function.BiFunction
import java.util.function.Supplier

interface HasBarRenderOnCreature_Power : HasBarRenderOnCreature {
    val self: AbstractPower
        get() = this as AbstractPower

    override fun makeNewBarRenderOnCreature():
            BiFunction<Supplier<Hitbox>, HasBarRenderOnCreature, out RenderOnThing> {
        return BiFunction<Supplier<Hitbox>, HasBarRenderOnCreature, RenderOnThing>(::BarRenderOnThing)
    }

    override fun uuidOfSelf(): String {
        return self.ID
    }

    override fun uuidPointTo(): String {
        if (self.owner is AbstractPlayer)
            return super.uuidPointTo() + ":player:" + self.owner.name
        return super.uuidPointTo() + ":" + self.owner.name
    }

    override fun getAmountForDraw(): Int = self.amount

    override fun maxBarAmount(): Int

    override val name: String
        get() = self.name

    override val description: String
        get() = self.description

    override fun getBarRenderHitBox(): Hitbox = self.owner.hb
}
